### PR TITLE
[RLlib] APPO (new API stack) enhancements vol 05: Auto-sleep time AND thread-safety for MetricsLogger.

### PR DESCRIPTION
**Pull request by sven1977**
APPO (new API stack) enhancements vol 05: Auto-sleep time AND thread-safety for MetricsLogger.

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: rllib/algorithms/impala/impala_learner.py [hunk 1]
```diff
@@ -324,26 +308,13 @@ def step(self):
...
-
-            self.metrics.log_value(
-                (ALL_MODULES, QUEUE_SIZE_RESULTS_QUEUE),
-                self._out_queue.qsize(),
+                timesteps=_CURRENT_GLOBAL_TIMESTEPS,
             )
 
     @staticmethod
-    def enqueue(learner_queue: deque, batch, metrics_logger):
+    def enqueue(learner_queue: deque, batch, metrics):
```
Comment (sven1977) [inline rllib/algorithms/impala/impala_learner.py:317]: Unify name (everywhere else, it's also called just `metrics`).

File: rllib/utils/metrics/stats.py [hunk 2]
```diff
@@ -295,11 +313,23 @@ class for details on the reduction logic applied to the values list, based on
...
 
+        # Keep track and update underlying throughput metric.
+        if self._throughput is not False:
+            # Take the delta between the new (upcoming) reduced value and the most
+            # recently reduced value (one `reduce()` call ago).
+            delta_sum = reduced - self._hist[-1]
+            assert delta_sum >= 0
+            time_now = time.perf_counter()
+            delta_time = time_now - self._throughput_last_time
+            assert delta_time >= 0.0
```
Comment (simonsays1980) [inline rllib/utils/metrics/stats.py:324]: Dumb question: Why do we need then still timing metrics? Couldn't we just log `num_env_steps_sampled_lifetime` and use a throughput?

**Review by simonsays1980 — APPROVED** [review]
LGTM. Big one. Left some questions.

File: rllib/core/learner/learner.py [hunk 3]
```diff
@@ -994,13 +994,14 @@ def update_from_batch(
                 new="Learner.update_from_episodes(num_epochs=...)",
                 error=True,
             )
-        return self._update_from_batch_or_episodes(
+        self._update_from_batch_or_episodes(
```
[the remark this answers is not in the document]
Comment (sven1977) [reply rllib/core/learner/learner.py:997]: Good question! Inside a single Learner.

File: rllib/core/learner/torch/torch_learner.py [hunk 4]
```diff
@@ -147,6 +147,13 @@ def _uncompiled_update(
         # Activate tensor-mode on our MetricsLogger.
         self.metrics.activate_tensor_mode()
 
+        # TODO (sven): Causes weird cuda error when WandB is used.
+        #  Diagnosis thus far:
+        #  - All peek values during metrics.reduce are non-tensors.
+        #  - However, in impala.py::training_step(), a tensor does arrive after learner
+        #    group.update_from_episodes(), so somehow, there is still a race condition
+        #    possible (learner, which performs the reduce() and learner thread, which
+        #    performs the logging of tensors into metrics logger).
```
[the remark this answers is not in the document]
Comment (sven1977) [reply rllib/core/learner/torch/torch_learner.py:156]: Hah, this seems to be resolved. Will remove the comment ...

File: rllib/algorithms/impala/impala.py [hunk 5]
```diff
@@ -875,6 +887,67 @@ def _process_data(_actor, _episodes):
...
+                "_measured_last_train_throughput",
+                self._last_train_throughput,
+                window=1,
+            )
+            # Improvement: Keep moving in the same direction with the same speed.
+            if train_throughput > self._last_train_throughput:
+                pass
+            # Got worse: Reverse direction and reduce learning rate.
+            else:
+                self._sleep_time_lr *= 0.95
```
[the remark this answers is not in the document]
Comment (sven1977) [reply rllib/algorithms/impala/impala.py:931]: This is a sleep-time specific learning rate. Has nothing to do with the actual learning rate (and is hence hidden from user). I don't want users to even think much about this setup. It should just work (and find the correct sleep time in a few minutes).

File: rllib/utils/metrics/metrics_logger.py [hunk 6]
```diff
@@ -884,21 +904,27 @@ def _reduce(path, stats):
...
-        # new empty `Stats` object from `stat.reduce()` with the same settings as
-        # existing one and can now re-assign it to `self.stats[key]` (while we return
-        # from this method the properly reduced, but not cleared/emptied new `Stats`).
+        # Create a shallow (yet nested) copy of `self.stats` in case we need to reset
+        # some of our stats due to this `reduce()` call and Stats having
+        # `self.clear_on_reduce=True`. In the latter case we would receive a new empty
+        # `Stats` object from `stat.reduce()` with the same settings as existing one and
+        # can now re-assign it to `self.stats[key]`, while we return from this method
+        # the properly reduced, but not cleared/emptied new `Stats`.
+        if key is not None:
```
[the remark this answers is not in the document]
Comment (sven1977) [reply rllib/utils/metrics/metrics_logger.py:913]: Correct! Users should never call `reduce()` anywhere on any MetricsLogger object. These calls sit only at the "handover" locations in RLlib's core code. Users simply have to log to the `MetricsLogger` and not worry about the underlying mechanics.

File: rllib/utils/metrics/metrics_logger.py [hunk 7]
```diff
@@ -997,10 +1026,19 @@ def set_value(
                 in which the internal values list would otherwise grow indefinitely,
                 for example if reduce is None and there is no `window` provided.
                 Note that this is only applied if `key` does not exist in `self` yet.
+            with_throughput: Whether to track a throughput estimate together with this
+                metric. This is only supported for `reduce=sum` and
+                `clear_on_reduce=False` metrics (aka. "lifetime counts"). The `Stats`
+                object under the logged key then keeps track of the time passed
+                between two consecutive calls to `reduce()` and update its throughput
+                estimate. The current throughput estimate of a key can be obtained
+                through: `MetricsLogger.peek([some key], throughput=True)`.
```
[the remark this answers is not in the document]
Comment (sven1977) [reply rllib/utils/metrics/metrics_logger.py:1035]: Only the throughput, if `throughput=True`, otherwise only the reduced value (no throughput).
